### ISSUE TITLE
What does the bad epochs mean？

### ISSUE THREAD

**Issue by aslicedbread** [zalandoresearch/flair#609]
Hi~
I found several bad epochs(like bad epochs 3) when browsing the training log. I‘m confused about the meaning of the bad epochs.’
One more question, if we want to use the bert-lstm-crf whose lstm layer is not bidirectional( I set the bidirectional = False in Sequence_tagger.py), but it seems to be more modified？(the error proves that dimension does not match)
Hope to get your reply, Thanks.

**Comment by gccome** [comment]
Hi, bad epoch means your metrics does not improve on validation set for this epoch

**Comment by aslicedbread** [comment]
> Hi, bad epoch means your metrics does not improve on validation set for this epoch

@gccome 
Hi,
What does the 3 mean? Does it mean that 3 epochs have not been improved on the verification set?

**Comment by gccome** [comment]
That’s correct.

**Comment by aslicedbread** [comment]
> That’s correct.

@gccome 
Thank you. 
And how does flair define the best model?
I mean that the highest F1 on the verification set throughout the training process is 0.735, but the F1 on the best model is 0.71. (the verification set and the testing set are the same. ) I'm also confused about this.

**Comment by gccome** [comment]
@aslicedbread by reviewing the code, the best model is saved when you have best F1 on validation set. At the end, the best model will be applied on test data. In my experiments, those two numbers match. Can I see you training log file?

**Comment by aslicedbread** [comment]
> @aslicedbread by reviewing the code, the best model is saved when you have best F1 on validation set. At the end, the best model will be applied on test data. In my experiments, those two numbers match. Can I see you training log file?
@gccome 
of course. Here is the url: https://share.weiyun.com/5U1NBWR
I'm not sure whether this web drive is available in your country.

**Comment by gccome** [comment]
@aslicedbread unfortunately I was not able to open it:(

**Comment by gatarelib** [comment]
> > @aslicedbread by reviewing the code, the best model is saved when you have best F1 on validation set. At the end, the best model will be applied on test data. In my experiments, those two numbers match. Can I see you training log file?
> > @gccome
> > of course. Here is the url: https://share.weiyun.com/5U1NBWR
> > I'm not sure whether this web drive is available in your country.

Unfortunately no.

**Comment by aslicedbread** [comment]
> @aslicedbread unfortunately I was not able to open it:(

@gccome 
sorry to hear that.
I have uploaded to the google drive.
Here is the url: 
https://drive.google.com/file/d/1dnlOaIrFqO4hVWyqNRl-C7PDPF1L59t3/view?usp=sharing

**Comment by gccome** [comment]
@aslicedbread it seems you were not evaluating again validation set. Could you please share your training code snippet?

**Comment by aslicedbread** [comment]
> @aslicedbread it seems you were not evaluating against validation set. Could you please share your training code snippet? Did you set `anneal_against_train_loss` to be False? By observing the log, seems like you were training against training loss.

@gccome 
of course.
Here is the url: 
https://drive.google.com/open?id=1niO_26tdOtV_tjdKdRA0bBSW1CXJtD3P
By the way, what does the `anneal_against_train_loss`  mean?

**Comment by gccome** [comment]
@aslicedbread Yes you were training against training loss, not validation set. If you look at `ModelTrainer.train` function, you will see that the default value for `anneal_against_train_loss` is `True`, meaning the best model is saved when you hit lowest training loss. You are not training against validation set unless you set this parameter to be `False`

**Comment by aslicedbread** [comment]
> @aslicedbread Yes you were training against training loss, not validation set. If you look at `ModelTrainer.train` function, you will see that the default value for `anneal_against_train_loss` is `True`, meaning the best model is saved when you hit lowest training loss. You are not training against validation set unless you set this parameter to be `False`

@gccome 
Thanks a lot. I get it.
In addition, what should I set if we want to use the bert-lstm-crf model (not the bidirectional lstm)?

**Comment by gccome** [comment]
@aslicedbread what do you mean by bert-lstm-crf? In flair you can use Bert embedding + lstm + crf for sequence tagging. Did you mean fine tuning BERT?

**Comment by aslicedbread** [comment]
> @aslicedbread what do you mean by bert-lstm-crf? In flair you can use Bert embedding + lstm + crf for sequence tagging. Did you mean fine tuning BERT?

@gccome 
No. I mean that we use the bert-bidirectional lstm-crf by default. If we want to use the bert-lstm(single direction)-crf, what should we set? ( I set the bidirectional = False in Sequence_tagger.py, but it seems that it's not enough. )

**Comment by gccome** [comment]
@aslicedbread I think set bidirectional=False should do. Why do you think it is not enough?

**Comment by alanakbik** [comment]
Closing due to inactivity, but feel free to reopen if there are more questions!

**Comment by oluwayetty** [comment]
![image](https://user-images.githubusercontent.com/10073272/92966586-1a235f00-f478-11ea-91f8-ef236a973753.png)
Could someone kindly explain why it's BAD EPOCHS despite that the loss reduced and score increase on the validation set?

**Comment by djstrong** [comment]
It is the bad epoch if the number is greater than 0.

**Comment by oluwayetty** [comment]
Thanks @djstrong If I understand you correctly, do you mean this doesn't mean Bad epochs? It's only bad if it's something like ```BAD EPOCKS (no improvement): 1``` or any number above zero?

**Comment by djstrong** [comment]
Yes, it is a counter of bad epochs.

**Comment by LangLoffelLako** [comment]
Hi,
I am someone buffled by the `BAD EPOCHS (no improvement)` counter as well. During training it shows 4, no matter how many epochs I train. Also the counter is at 4 already after the first epoch, which seems counterintuitive.
![Bad_epochs_bug](https://user-images.githubusercontent.com/49274015/196976685-e48881f9-a658-4538-871b-186a4ab2c7a9.PNG)
Can someone explain what I am missing here?

**Comment by alanakbik** [comment]
Hello @LangLoffelLako, the BAD EPOCHS printout only makes sense if used in combination with annealing. This is default in training, but likely you are doing fine-tuning in which case you can ignore the printout. I'll put in a PR to remove this since I agree it is confusing.
